File: design-pattern-php/Factory/SimpleFactory.php
<?php

/**
 * 简单/静态工厂
 */

require_once __DIR__ . "/Animal.php";

class AnimalFactory
{
    /**
     * 制造动物
     *
     * @param   string  $type  动物类型
     *
     * @return  Animal
     */
    public static function make(string $type): Animal
    {
        switch ($type) {
            case "Cat":
            case "cat":
                return new Cat;
                break;
            case "Dog":
            case "dog":
                return new Dog;
                break;
            default:
                return new $type;
        }
    }
}

$cat = AnimalFactory::make('cat');
$dog = AnimalFactory::make('dog');

var_dump($cat, $dog);

// F:\Code\code\design-pattern\Factory\SimpleFactory.php:49:
// class Cat#1 (1) {
//   public $name =>
//   string(3) "Cat"
// }
// F:\Code\code\design-pattern\Factory\SimpleFactory.php:49:
// class Dog#2 (1) {
//   public $name =>
//   string(3) "Dog"
// }

File: design-pattern-java/src/main/java/me/ixk/design_pattern/factory/SimpleFactory.java
package me.ixk.design_pattern.factory;

import me.ixk.design_pattern.factory.animal.Animal;
import me.ixk.design_pattern.factory.animal.Cat;
import me.ixk.design_pattern.factory.animal.Dog;

/**
 * 简单工厂
 * <p>
 * 优点：简单易用
 * <p>
 * 缺点：当工厂需要制作新的类型的实例的时候需要修改工厂的代码
 */
public class SimpleFactory {

    public <T extends Animal> T makeAnimal(final Class<T> animalType) {
        final Animal animal;
        if (animalType == Cat.class) {
            animal = new Cat();
        } else {
            animal = new Dog();
        }
        return animalType.cast(animal);
    }
}
